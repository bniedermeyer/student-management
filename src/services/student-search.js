const searchForStudents = async filters => {
  if (filters) {
    console.log('performing search with following filters', filters);
  }

  const result = await fetch(
    'https://my-json-server.typicode.com/bniedermeyer/student-management/students'
  );
  return result.json();
};

module.exports = {
  searchForStudents
};
