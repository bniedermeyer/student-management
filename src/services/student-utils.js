/**
 * Queries the mock json server for entries that match the givern filters.
 * @param  filters
 */
const searchForStudents = async filters => {
  // todo: add paging for search - since this is a demo project it's ok. In a larger app you'd need a way to not be drowned by the load of results
  let url =
    'https://my-json-server.typicode.com/bniedermeyer/student-management/students';

  if (filters) {
    let filtersString = '';
    for (const key in filters) {
      if (filters.hasOwnProperty(key)) {
        const value = filters[key];

        if (value.length > 0) {
          if (filtersString.length > 0) {
            filtersString += '&';
          }
          filtersString += `${key}_like=${value}`;
        }
      }
    }

    if (filtersString.length > 0) {
      url += `?${filtersString}`;
    }
  }

  const result = await fetch(url);
  return await result.json();
};

/**
 * sends a POST request to the json server to add the student to the database.
 * Note - because the project uses my-json-server to host the mock DB, the response is faked by
 * the server and not persisted
 * @param  student
 */
const saveStudentToDatabase = async student => {
  let url =
    'https://my-json-server.typicode.com/bniedermeyer/student-management/students';

  const response = await fetch(url, {
    method: 'POST',
    body: JSON.stringify(student)
  });

  return await response.json();
};

export { searchForStudents, saveStudentToDatabase };
