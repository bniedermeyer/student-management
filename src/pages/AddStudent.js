import React, { useState } from 'react';
import {
  Pane,
  Heading,
  Card,
  Button,
  TextInputField,
  minorScale,
  RadioGroup,
  toaster
} from 'evergreen-ui';

import { addStudent } from '../services/student-utils';

const AddStudent = () => {
  const defaultStudent = {
    name: '',
    active: 'true',
    homeCountry: '',
    exchangeCountry: '',
    homeUni: '',
    gpa: ''
  };
  const [student, setStudent] = useState(defaultStudent);

  const updateStudent = (field, value) => {
    const updatedStudent = { ...student, [field]: value };
    setStudent(updatedStudent);
  };

  const saveStudent = async e => {
    e.preventDefault();
    await addStudent(student);
    toaster.success('Student Created');
  };

  return (
    <Pane>
      <Heading size={900} is="h1">
        Add a Student
      </Heading>
      <Card
        elevation={3}
        border
        padding={minorScale(8)}
        width="50%"
        marginLeft="25%"
        marginTop={minorScale(4)}
        textAlign="left"
      >
        <form onSubmit={e => saveStudent(e)}>
          <Pane display="grid" gridTemplateColumns="repeat(2, auto)">
            <TextInputField
              label="Name"
              value={student.name}
              marginRight={minorScale(2)}
              onChange={e => updateStudent('name', e.target.value)}
            />
            <TextInputField
              label="Home Country"
              value={student.homeCountry}
              marginRight={minorScale(2)}
              onChange={e => updateStudent('homeCountry', e.target.value)}
            />
            <TextInputField
              label="Home University"
              value={student.homeUni}
              marginRight={minorScale(2)}
              onChange={e => updateStudent('homeUni', e.target.value)}
            />
            <TextInputField
              label="ExchangeCountry"
              value={student.exchangeCountry}
              marginRight={minorScale(2)}
              onChange={e => updateStudent('exchangeCountry', e.target.value)}
            />
            <TextInputField
              label="GPA"
              value={student.gpa}
              marginRight={minorScale(2)}
              onChange={e => updateStudent('gpa', e.target.value)}
            />
            <RadioGroup
              label="Active"
              value={student.active}
              options={[
                { label: 'Active', value: 'true' },
                { label: 'Inactive', value: 'false' }
              ]}
              onChange={value => updateStudent('active', value)}
            />
          </Pane>
          <Pane display="flex" justifyContent="flex-end">
            <Button
              type="button"
              onClick={() => setStudent(defaultStudent)}
              marginRight={minorScale(2)}
            >
              Reset Form
            </Button>
            <Button
              marginRight={16}
              appearance="primary"
              iconBefore="save"
              type="submit"
            >
              Save
            </Button>
          </Pane>
        </form>
      </Card>
    </Pane>
  );
};

export default AddStudent;
