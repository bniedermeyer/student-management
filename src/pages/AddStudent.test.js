import React from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import renderer, { act } from 'react-test-renderer';
import AddStudent from './AddStudent';
import { saveStudentToDatabase } from '../services/student-utils';

jest.mock('../services/student-utils.js', () => {
  return {
    __esModule: true,
    saveStudentToDatabase: jest.fn(() => Promise.resolve({ id: 1 }))
  };
});

let container = null;
beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

it('should render the page', () => {
  const tree = renderer.create(<AddStudent />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should save the student to the database', async () => {
  act(() => {
    render(<AddStudent />, container);
  });
  const button = container.querySelector('.btn-save');
  const name = container.querySelector('.student-name');

  act(() => {
    var nativeInputValueSetter = Object.getOwnPropertyDescriptor(
      window.HTMLInputElement.prototype,
      'value'
    ).set;
    nativeInputValueSetter.call(name, 'Ron Swanson');

    const event = new Event('change', { bubbles: true });
    name.dispatchEvent(event);
    button.dispatchEvent(new MouseEvent('click', { bubbles: true }));
  });

  expect(saveStudentToDatabase).toBeCalledWith({
    name: 'Ron Swanson',
    active: 'true',
    homeCountry: '',
    exchangeCountry: '',
    homeUni: '',
    gpa: ''
  });
});
