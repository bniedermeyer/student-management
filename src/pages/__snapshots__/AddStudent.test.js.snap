// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the page 1`] = `
<div
  className="css-nil ðŸ“¦box-szg_border-box"
>
  <h1
    className="ðŸ“¦mt_0px ðŸ“¦mb_0px ðŸ“¦fnt-sze_35px ðŸ“¦f-wght_500 ðŸ“¦ln-ht_40px ðŸ“¦ltr-spc_-0-2px ðŸ“¦fnt-fam_wm9v6k ðŸ“¦color_234361 ðŸ“¦box-szg_border-box"
  >
    Add a Student
  </h1>
  <div
    className="css-nil ðŸ“¦b-top_1px-solid-E4E7EB ðŸ“¦b-rgt_1px-solid-E4E7EB ðŸ“¦b-btm_1px-solid-E4E7EB ðŸ“¦b-lft_1px-solid-E4E7EB ðŸ“¦bs_17sc08g ðŸ“¦bblr_5px ðŸ“¦bbrr_5px ðŸ“¦btlr_5px ðŸ“¦btrr_5px ðŸ“¦pb_32px ðŸ“¦pl_32px ðŸ“¦pr_32px ðŸ“¦pt_32px ðŸ“¦w_50prcnt ðŸ“¦ml_25prcnt ðŸ“¦mt_16px ðŸ“¦txt-algn_left ðŸ“¦box-szg_border-box"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="css-nil ðŸ“¦dspl_grid ðŸ“¦grd-tmp-col_fefeyf ðŸ“¦box-szg_border-box"
      >
        <div
          className="ðŸ“¦mb_24px ðŸ“¦mr_8px ðŸ“¦box-szg_border-box"
        >
          <label
            className="ðŸ“¦color_425A70 ðŸ“¦fnt-fam_b77syt ðŸ“¦fnt-sze_14px ðŸ“¦f-wght_500 ðŸ“¦ln-ht_20px ðŸ“¦ltr-spc_-0-05px ðŸ“¦dspl_block ðŸ“¦mb_4px ðŸ“¦box-szg_border-box"
            htmlFor="TextInputField-0"
          >
            Name
             
          </label>
          <input
            aria-invalid={false}
            className="css-5ljhhe student-name ðŸ“¦color_425A70 ðŸ“¦fnt-fam_b77syt ðŸ“¦fnt-sze_12px ðŸ“¦f-wght_400 ðŸ“¦ln-ht_16px ðŸ“¦ltr-spc_0 ðŸ“¦w_100prcnt ðŸ“¦h_32px ðŸ“¦pl_10px ðŸ“¦pr_10px ðŸ“¦bblr_3px ðŸ“¦bbrr_3px ðŸ“¦btlr_3px ðŸ“¦btrr_3px ðŸ“¦box-szg_border-box"
            disabled={false}
            id="TextInputField-0"
            onChange={[Function]}
            spellCheck={true}
            type="text"
            value=""
          />
        </div>
        <div
          className="ðŸ“¦mb_24px ðŸ“¦mr_8px ðŸ“¦box-szg_border-box"
        >
          <label
            className="ðŸ“¦color_425A70 ðŸ“¦fnt-fam_b77syt ðŸ“¦fnt-sze_14px ðŸ“¦f-wght_500 ðŸ“¦ln-ht_20px ðŸ“¦ltr-spc_-0-05px ðŸ“¦dspl_block ðŸ“¦mb_4px ðŸ“¦box-szg_border-box"
            htmlFor="TextInputField-1"
          >
            Home Country
             
          </label>
          <input
            aria-invalid={false}
            className="css-5ljhhe home-country ðŸ“¦color_425A70 ðŸ“¦fnt-fam_b77syt ðŸ“¦fnt-sze_12px ðŸ“¦f-wght_400 ðŸ“¦ln-ht_16px ðŸ“¦ltr-spc_0 ðŸ“¦w_100prcnt ðŸ“¦h_32px ðŸ“¦pl_10px ðŸ“¦pr_10px ðŸ“¦bblr_3px ðŸ“¦bbrr_3px ðŸ“¦btlr_3px ðŸ“¦btrr_3px ðŸ“¦box-szg_border-box"
            disabled={false}
            id="TextInputField-1"
            onChange={[Function]}
            spellCheck={true}
            type="text"
            value=""
          />
        </div>
        <div
          className="ðŸ“¦mb_24px ðŸ“¦mr_8px ðŸ“¦box-szg_border-box"
        >
          <label
            className="ðŸ“¦color_425A70 ðŸ“¦fnt-fam_b77syt ðŸ“¦fnt-sze_14px ðŸ“¦f-wght_500 ðŸ“¦ln-ht_20px ðŸ“¦ltr-spc_-0-05px ðŸ“¦dspl_block ðŸ“¦mb_4px ðŸ“¦box-szg_border-box"
            htmlFor="TextInputField-2"
          >
            Home University
             
          </label>
          <input
            aria-invalid={false}
            className="css-5ljhhe home-university ðŸ“¦color_425A70 ðŸ“¦fnt-fam_b77syt ðŸ“¦fnt-sze_12px ðŸ“¦f-wght_400 ðŸ“¦ln-ht_16px ðŸ“¦ltr-spc_0 ðŸ“¦w_100prcnt ðŸ“¦h_32px ðŸ“¦pl_10px ðŸ“¦pr_10px ðŸ“¦bblr_3px ðŸ“¦bbrr_3px ðŸ“¦btlr_3px ðŸ“¦btrr_3px ðŸ“¦box-szg_border-box"
            disabled={false}
            id="TextInputField-2"
            onChange={[Function]}
            spellCheck={true}
            type="text"
            value=""
          />
        </div>
        <div
          className="ðŸ“¦mb_24px ðŸ“¦mr_8px ðŸ“¦box-szg_border-box"
        >
          <label
            className="ðŸ“¦color_425A70 ðŸ“¦fnt-fam_b77syt ðŸ“¦fnt-sze_14px ðŸ“¦f-wght_500 ðŸ“¦ln-ht_20px ðŸ“¦ltr-spc_-0-05px ðŸ“¦dspl_block ðŸ“¦mb_4px ðŸ“¦box-szg_border-box"
            htmlFor="TextInputField-3"
          >
            ExchangeCountry
             
          </label>
          <input
            aria-invalid={false}
            className="css-5ljhhe exchange-country ðŸ“¦color_425A70 ðŸ“¦fnt-fam_b77syt ðŸ“¦fnt-sze_12px ðŸ“¦f-wght_400 ðŸ“¦ln-ht_16px ðŸ“¦ltr-spc_0 ðŸ“¦w_100prcnt ðŸ“¦h_32px ðŸ“¦pl_10px ðŸ“¦pr_10px ðŸ“¦bblr_3px ðŸ“¦bbrr_3px ðŸ“¦btlr_3px ðŸ“¦btrr_3px ðŸ“¦box-szg_border-box"
            disabled={false}
            id="TextInputField-3"
            onChange={[Function]}
            spellCheck={true}
            type="text"
            value=""
          />
        </div>
        <div
          className="ðŸ“¦mb_24px ðŸ“¦mr_8px ðŸ“¦box-szg_border-box"
        >
          <label
            className="ðŸ“¦color_425A70 ðŸ“¦fnt-fam_b77syt ðŸ“¦fnt-sze_14px ðŸ“¦f-wght_500 ðŸ“¦ln-ht_20px ðŸ“¦ltr-spc_-0-05px ðŸ“¦dspl_block ðŸ“¦mb_4px ðŸ“¦box-szg_border-box"
            htmlFor="TextInputField-4"
          >
            GPA
             
          </label>
          <input
            aria-invalid={false}
            className="css-5ljhhe gpa ðŸ“¦color_425A70 ðŸ“¦fnt-fam_b77syt ðŸ“¦fnt-sze_12px ðŸ“¦f-wght_400 ðŸ“¦ln-ht_16px ðŸ“¦ltr-spc_0 ðŸ“¦w_100prcnt ðŸ“¦h_32px ðŸ“¦pl_10px ðŸ“¦pr_10px ðŸ“¦bblr_3px ðŸ“¦bbrr_3px ðŸ“¦btlr_3px ðŸ“¦btrr_3px ðŸ“¦box-szg_border-box"
            disabled={false}
            id="TextInputField-4"
            onChange={[Function]}
            spellCheck={true}
            type="text"
            value=""
          />
        </div>
        <div
          aria-label="Active"
          className="active css-nil ðŸ“¦box-szg_border-box"
          role="group"
        >
          <span
            className="ðŸ“¦color_66788A ðŸ“¦fnt-fam_b77syt ðŸ“¦fnt-sze_14px ðŸ“¦f-wght_500 ðŸ“¦ln-ht_20px ðŸ“¦ltr-spc_-0-05px ðŸ“¦box-szg_border-box"
          >
            Active
          </span>
          <label
            className="ðŸ“¦crsr_pointer ðŸ“¦pst_relative ðŸ“¦dspl_flex ðŸ“¦mb_8px ðŸ“¦mt_8px ðŸ“¦box-szg_border-box"
          >
            <input
              aria-invalid={false}
              checked={true}
              className="css-1bkiqpw ðŸ“¦box-szg_border-box"
              name="RadioGroup-1"
              onChange={[Function]}
              required={false}
              type="radio"
              value="true"
            />
            <div
              className="ðŸ“¦box-szg_border-box ðŸ“¦bblr_9999px ðŸ“¦bbrr_9999px ðŸ“¦btlr_9999px ðŸ“¦btrr_9999px ðŸ“¦dspl_flex ðŸ“¦flx_none ðŸ“¦algn-itms_center ðŸ“¦just-cnt_center ðŸ“¦mt_2px ðŸ“¦w_12px ðŸ“¦h_12px"
            >
              <svg
                height={4}
                viewBox="0 0 10 10"
                width={4}
              >
                <circle
                  cx="5"
                  cy="5"
                  fill="currentColor"
                  r="5"
                />
              </svg>
            </div>
            <span
              className="ðŸ“¦color_425A70 ðŸ“¦fnt-fam_b77syt ðŸ“¦fnt-sze_12px ðŸ“¦f-wght_400 ðŸ“¦ln-ht_16px ðŸ“¦ltr-spc_0 ðŸ“¦ml_8px ðŸ“¦box-szg_border-box"
            >
              Active
            </span>
          </label>
          <label
            className="ðŸ“¦crsr_pointer ðŸ“¦pst_relative ðŸ“¦dspl_flex ðŸ“¦mb_8px ðŸ“¦mt_8px ðŸ“¦box-szg_border-box"
          >
            <input
              aria-invalid={false}
              checked={false}
              className="css-1bkiqpw ðŸ“¦box-szg_border-box"
              name="RadioGroup-1"
              onChange={[Function]}
              required={false}
              type="radio"
              value="false"
            />
            <div
              className="ðŸ“¦box-szg_border-box ðŸ“¦bblr_9999px ðŸ“¦bbrr_9999px ðŸ“¦btlr_9999px ðŸ“¦btrr_9999px ðŸ“¦dspl_flex ðŸ“¦flx_none ðŸ“¦algn-itms_center ðŸ“¦just-cnt_center ðŸ“¦mt_2px ðŸ“¦w_12px ðŸ“¦h_12px"
            >
              <svg
                height={4}
                viewBox="0 0 10 10"
                width={4}
              >
                <circle
                  cx="5"
                  cy="5"
                  fill="currentColor"
                  r="5"
                />
              </svg>
            </div>
            <span
              className="ðŸ“¦color_425A70 ðŸ“¦fnt-fam_b77syt ðŸ“¦fnt-sze_12px ðŸ“¦f-wght_400 ðŸ“¦ln-ht_16px ðŸ“¦ltr-spc_0 ðŸ“¦ml_8px ðŸ“¦box-szg_border-box"
            >
              Inactive
            </span>
          </label>
        </div>
      </div>
      <div
        className="css-nil ðŸ“¦dspl_flex ðŸ“¦just-cnt_flex-end ðŸ“¦box-szg_border-box"
      >
        <button
          className="css-69cngj btn-reset ðŸ“¦fnt-fam_b77syt ðŸ“¦mt_0px ðŸ“¦fnt-sze_12px ðŸ“¦f-wght_500 ðŸ“¦ln-ht_32px ðŸ“¦ltr-spc_0 ðŸ“¦btrr_3px ðŸ“¦bbrr_3px ðŸ“¦btlr_3px ðŸ“¦bblr_3px ðŸ“¦pt_0px ðŸ“¦pb_0px ðŸ“¦pr_16px ðŸ“¦pl_16px ðŸ“¦ml_0px ðŸ“¦mr_8px ðŸ“¦mb_0px ðŸ“¦h_32px ðŸ“¦pst_relative ðŸ“¦dspl_inline-flex ðŸ“¦algn-itms_center ðŸ“¦flx-wrap_nowrap ðŸ“¦box-szg_border-box"
          onClick={[Function]}
          type="button"
        >
          Reset Form
        </button>
        <button
          className="css-nwvixr btn-save ðŸ“¦fnt-fam_b77syt ðŸ“¦mt_0px ðŸ“¦fnt-sze_12px ðŸ“¦f-wght_500 ðŸ“¦ln-ht_32px ðŸ“¦ltr-spc_0 ðŸ“¦btrr_3px ðŸ“¦bbrr_3px ðŸ“¦btlr_3px ðŸ“¦bblr_3px ðŸ“¦pt_0px ðŸ“¦pb_0px ðŸ“¦pr_16px ðŸ“¦pl_16px ðŸ“¦ml_0px ðŸ“¦mr_16px ðŸ“¦mb_0px ðŸ“¦h_32px ðŸ“¦pst_relative ðŸ“¦dspl_inline-flex ðŸ“¦algn-itms_center ðŸ“¦flx-wrap_nowrap ðŸ“¦box-szg_border-box"
          type="submit"
        >
          Save
        </button>
      </div>
    </form>
  </div>
</div>
`;
